local-cloudbuild.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python
+#
+# Run the steps of cloudbuild.yaml locally, emulating Google Container
+# Builder functionality.  Does not actually push images to the
+# Container Registry.  Not all functionality is supported.
+#
+# Based on https://cloud.google.com/container-builder/docs/api/build-steps
+
+# System packages
+import argparse
+import getpass
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+# Third party packages
+import yaml
+
+
+# Types
+class CloudBuildError(Exception):
+    pass
+
+
+def main(argv):
+    """Main entrypoint for cli"""
+    parser = argparse.ArgumentParser(
```

Comment:
Actually, if we're talking about testability I specifically recommend that for testability. If `main` requires `argparse`, then you either have to pass in `argv` to the `main` in your tests or mock `sys.argv`. Either way, it's not explicit. Better to do this:

```python

def main(arg1, arg2, arg3):
    ...

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('arg1')
    ...
    args = parser.parse_args()
    main(args.arg1, args.arg2, args.arg3)

```

Your tests can then test `main` like any other function with explicit arguments:

```python

def test_main():
    main(arg1, arg2, arg3)
```

If you wish to also test the argparse section, then just create a separate function called `parse_args` that contains all of that and returns `args`, then change your `if __name__` to be:

```
args = parse_args()
main(args.arg1, args.arg2, ...)
```

This again allows you to test argparsing without any mocking.